### PR TITLE
External process interface

### LINKED ISSUE #1953: case_run do_external masks output
output from the external script is not printed.

### PR DESCRIPTION
**Pull request by jedwards4b**
Improve interface to external scripts PRERUN_SCRIPT, POSTRUN_SCRIPT, DATA_ASSIMILATION_SCRIPT.   If DATA_ASSIMILATION is true but no script is named throw an error, allow scripts in python to be called directly instead of through shell.   Improve handling of arguments and log files. 

Test suite: scripts_regression_tests.py, hand testing
Test baseline: 
Test namelist changes: 
Test status: bit for bit
Fixes #1953

User interface changes?: 

Update gh-pages html (Y/N)?: Y

Code review:

### CODE REVIEW

**Comment by billsacks** [comment]
@JeremyFyke @lofverstrom - this may be relevant for your plugging in of the topo-updating script.

**Comment by JeremyFyke** [comment]
Thanks @billsacks - agreed.

**Comment by jedwards4b** [comment]
@gold2718 @kdraeder - If you can't review in a timely manner, please take yourself off the list of reviewers and I will find someone else.

File: doc/source/users_guide/running-a-case.rst [hunk 1]
```diff
@@ -457,4 +457,52 @@ long-term archiver tool that supported mass tape storage and HPSS systems.
 However, with the industry migration away from tape archives, it is no longer
 feasible for CIME to support all the possible archival schemes available.
 
+============================
+Data Assimilation and other External Processing
+============================
+
+Users may desire to run a pre or post run task on compute nodes or a data assimilation mechanism.
```
Comment (gold2718) [inline doc/source/users_guide/running-a-case.rst]: This wording feels a bit awkward to me. I suggest breaking it into two sentences:

> CIME provides the ability to run a task on compute nodes before or after a model run. CIME also provides a mechanism to run data assimilation steps as part of a submitted job (i.e., `./case.submit`).

File: doc/source/users_guide/running-a-case.rst [hunk 2]
```diff
@@ -457,4 +457,52 @@ long-term archiver tool that supported mass tape storage and HPSS systems.
...
+
+---------------------
+Pre and Post run scripts
+---------------------
+
+Variables ``PRERUN_SCRIPT`` and ``POSTRUN_SCRIPT`` can be used to name
```
Comment (gold2718) [inline doc/source/users_guide/running-a-case.rst]: "can be used to name" ==> "can each be used to name"

**Comment by jedwards4b** [comment]
@gold2718 I've updated the documentation with your suggestions.

**Comment by jgfouca** [comment]
@jedwards4b , I don't know how to do that off the top of my head. I could investigate if you want.

**Comment by jedwards4b** [comment]
I've opened issue #2029 to address the disposition of logs.
@jgfouca I think we can merge this one.

**Pull request merged**